diff --git a/core/jni/android_net_wifi_Wifi.cpp b/core/jni/android_net_wifi_Wifi.cpp
index d36ac4e..0a509b1 100644
--- a/core/jni/android_net_wifi_Wifi.cpp
+++ b/core/jni/android_net_wifi_Wifi.cpp
@@ -190,11 +190,6 @@ static jstring android_net_wifi_doStringCommand(JNIEnv* env, jobject, jstring jI
     return doStringCommand(env, ifname.c_str(), "%s", command.c_str());
 }
 
-static jboolean android_net_wifi_setMode(JNIEnv* env, jobject, jint type)
-{
-    return (jboolean)(::wifi_set_mode(type) == 0);
-}
-
 // ----------------------------------------------------------------------------
 
 /*
@@ -220,7 +215,6 @@ static JNINativeMethod gWifiMethods[] = {
             (void*) android_net_wifi_doIntCommand },
     { "doStringCommand", "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;",
             (void*) android_net_wifi_doStringCommand },
-    { "setMode", "(I)Z", (void*) android_net_wifi_setMode},
 };
 
 int register_android_net_wifi_WifiManager(JNIEnv* env)
