diff --git a/VolumeManager.cpp b/VolumeManager.cpp
index a944543..df419f3 100644
--- a/VolumeManager.cpp
+++ b/VolumeManager.cpp
@@ -1189,7 +1189,7 @@ static const char *LUN_FILES[] = {
 #endif
     /* Only andriod0 exists, but the %d in there is a hack to satisfy the
        format string and also give a not found error when %d > 0 */
-    "/sys/class/android_usb/android%d/f_mass_storage/lun/file",
+    "/sys/class/android_usb/android%d/f_mass_storage/lun0/file",
     NULL
 };
 
